### PR TITLE
Enforcing consistent access to environment variables

### LINKED ISSUE #1592: environmentSetupScript capitalizes environment variable names
### Brief Issue Summary

`environmentSetupScript` in `cmake-kits.json` converts all exported environment variable names to uppercase.

### Expected:

Variable names are preserved as is.

### Apparent Behavior:

Variable names are capitalized.

### CMake Tools Log

```
[main] Configuring folder: cmake-test 
[proc] Executing command: /usr/bin/cmake --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Release -H/home/aqtest/cmake-test -B/home/aqtest/cmake-test/build -G "Unix Makefiles"
[cmake] Not searching for unused variables given on the command line.
[cmake] LOWERCASEVAR=value
[cmake] -- Configuring done
[cmake] -- Generating done
[cmake] -- Build files have been written to: /home/aqtest/cmake-test/build
```

### Platform and Versions

- **Operating System**: Ubuntu
- **CMake Version**: 3.16.3
- **VSCode Version**: 1.51.1
- **CMake Tools Extension Version**: 1.5.3
- **Compiler/Toolchain**: Any

### Minimal project example to reproduce:

```
.vscode
|   cmake-kits.json
CMakeLists.txt
envsetup.sh
```

**cmake-kits.json**

```json
[
    {
        "name": "TestKit",
        "environmentSetupScript": "${workspaceFolder}/envsetup.sh"
    }
]
```

**CMakeLists.txt**

```cmake
cmake_minimum_required(VERSION 3.10)

project(TEST)

execute_process(COMMAND "bash" "-c" "env | grep -i lowercasevar")
```

**envsetup.sh**

```bash
#!/bin/bash
export lowercasevar=value
```

Steps:

1. Select kit as `TestKit`
2. Configure.

### LINKED ISSUE #2237: debugTarget failed when debug_config.environment not present.
debug_config.environment = debug_config.environment ? debug_config.environment.concat(util.makeDebuggerEnvironmentVars(configure_preset_environment)) : [];

means when debug_config have no environment  then  we can not debugging at all
even thout preset or toolkist are already configured it properly

### PR DESCRIPTION
**Pull request by lygstate**
## This change addresses item
Fixes #1592
Replacement of  #2025

Fixes #2237

~~Fixes https://github.com/microsoft/vscode-cmake-tools/issues/1987~~

### This changes

The following changes are proposed:
- Refactoring environmentVariables.ts out for consistence environment variable handling with NodeJS.ProcessEnv

## The purpose of this change
- Get environment variables on win32 properly, that is we can set/get environment variables in case-insensitive way
- Add backendTests to the CI, and revise the launch.json for faster debugging utils funcitons
- NOTE: Future direction, moving tests that's not depends on VSCode into backendTests

### CODE REVIEW

**Comment by bobbrow** [comment]
Thank you for continuing to work on this!  I'll try to take a look soon, but I currently have some vacation time planned and other time critical things to complete this month which may make that difficult.

**Comment by lygstate** [comment]
@bobbrow Hi, ping for review to avoid conflict with the repository

**Comment by bobbrow** [comment]
I will be looking at this today.

File: src/environmentVariables.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    toString() {
+        return this[envProperty].toString();
+    }
+}
+
+export class EnvironmentVariablesUtils {
+
+    public static create(from?: Map<string, string> | EnvironmentVariables | null, preserveNull?: boolean, isWin32?: boolean): EnvironmentVariablesUndefined {
+        const env = new EnvironmentVariablesPrivate(preserveNull, isWin32);
+        const p = new Proxy(env, {
```
Comment (bobbrow) [inline src/environmentVariables.ts:92]: A lot of complication in this class is because of this Proxy. Can you help me understand why we need to use a Proxy?  I was assuming it was to allow for indexing syntax, but that seems to be [supported in TypeScript](https://www.typescriptlang.org/docs/handbook/2/classes.html#index-signatures) already.

File: src/environmentVariables.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+        return p;
+    }
+
+    public static createPreserveNull(from?: Map<string, string> | EnvironmentVariables | null): EnvironmentVariables {
+        return EnvironmentVariablesUtils.create(from, true);
+    }
+
+    public static mergeImpl(preserveNull: boolean, isWin32: boolean | undefined, ...envs: EnvironmentVariablesItem[]): EnvironmentVariablesUndefined {
```
Comment (bobbrow) [inline src/environmentVariables.ts:129]: An `impl` method shouldn't be public.  Can we just have:
```ts
public static merge(options: EnvironmentOptions, ...envs: Environment[]): Environment
```
You can remove the other convenience methods.

File: src/environmentVariables.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                    p[key] = value;
+                }
+            } else {
+                Object.assign(p, from);
+            }
+        }
+        return p;
+    }
+
+    public static createPreserveNull(from?: Map<string, string> | EnvironmentVariables | null): EnvironmentVariables {
```
Comment (bobbrow) [inline src/environmentVariables.ts:125]: I would rather not have another function if you can already pass the information to `create` which is also public.

File: src/environmentVariables.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                return Reflect.get(target, p);
+            },
+            getOwnPropertyDescriptor: (target, p) => {
+                if (typeof p === 'string') {
+                    return Reflect.getOwnPropertyDescriptor(target[envProperty], target.getKey(p, false));
+                } else {
+                    return Reflect.getOwnPropertyDescriptor(target, p);
+                }
+            }
+            ,
```
Comment (bobbrow) [inline src/environmentVariables.ts:108]: comma should go up one line

File: src/environmentVariables.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    [inspect]() {
+        return util.inspect(this[envProperty]);
+    }
+
+    toString() {
+        return this[envProperty].toString();
+    }
+}
+
+export class EnvironmentVariablesUtils {
```
Comment (bobbrow) [inline src/environmentVariables.ts:88]: It would be really nice if `create` and `merge` could be on `Environment` as static methods.  Because of the proxy, we can't do this right now. If we can't remove the proxy, then please rename this `EnvironmentUtils` instead.

File: src/expand.ts [hunk 6]
```diff
@@ -95,9 +95,9 @@ export interface ExpansionOptions {
  * @param opts Options for the expansion process
  * @returns A string with the variable references replaced
  */
-export async function expandString(tmpl: string, opts: ExpansionOptions) {
+export async function expandString(tmpl: string | null | undefined, opts: ExpansionOptions): Promise<string> {
     if (!tmpl) {
-        return tmpl;
+        return tmpl as string;
```
Comment (bobbrow) [inline src/expand.ts:100]: `as` is not type coercion.  If we must return a string here, return `""` explicitly.

File: src/expand.ts [hunk 7]
```diff
@@ -121,8 +121,8 @@ export async function expandString(tmpl: string, opts: ExpansionOptions) {
 }
 
 async function expandStringHelper(tmpl: string, opts: ExpansionOptions) {
-    const envPreNormalize = opts.envOverride ? opts.envOverride : process.env as EnvironmentVariables;
-    const env = mergeEnvironment(envPreNormalize);
+    const envPreNormalize = opts.envOverride ? opts.envOverride : process.env;
+    const env = EnvironmentVariablesUtils.merge(envPreNormalize);
```
Comment (bobbrow) [inline src/expand.ts:125]: `create`?

File: src/kit.ts [hunk 8]
```diff
@@ -1108,32 +1108,31 @@ export async function getVSKitEnvironment(kit: Kit): Promise<Map<string, string>
     return varsForVSInstallation(requested, kit.visualStudioArchitecture!, kit.preferredGenerator?.platform);
 }
 
-export async function effectiveKitEnvironment(kit: Kit, opts?: expand.ExpansionOptions): Promise<Map<string, string>> {
-    let host_env;
-    const kit_env = objectPairs(kit.environmentVariables || {});
+export async function effectiveKitEnvironment(kit: Kit, opts?: expand.ExpansionOptions): Promise<EnvironmentVariablesUndefined> {
+    let host_env: EnvironmentVariablesUndefined | undefined;
+    const kit_env = EnvironmentVariablesUtils.merge(kit.environmentVariables);
```
Comment (bobbrow) [inline src/kit.ts:1113]: `create`?

File: src/preset.ts [hunk 9]
```diff
@@ -925,7 +926,7 @@ async function expandConfigurePresetHelper(folder: string,
             const compilerName: string | undefined = util.isSupportedCompiler(cxxCompiler) || util.isSupportedCompiler(cCompiler);
             if (compilerName) {
                 const compilerLocation = await execute('where.exe', [compilerName], null, {
-                    environment: preset.environment as EnvironmentVariables,
+                    environment: EnvironmentVariablesUtils.merge(preset.environment),
```
Comment (bobbrow) [inline src/preset.ts:929]: `create`

File: src/util.ts [hunk 10]
```diff
@@ -367,34 +371,30 @@ export function* flatMap<In, Out>(rng: Iterable<In>, fn: (item: In) => Iterable<
...
-            return { ...acc, ...norm_vars };
-        } else {
-            return { ...acc, ...vars };
+export function makeDebuggerEnvironmentVars(env?: EnvironmentVariablesUndefined): DebuggerEnvironmentVariable[] {
+    if (!env) {
+        return [];
+    }
+    const converted_env: DebuggerEnvironmentVariable[] = [];
+    for (const [key, value] of Object.entries(env)) {
+        if (value) {
```
Comment (bobbrow) [inline src/util.ts:380]: I think it should be an explicit check against null `value === null`.  We want to keep `0` and `""`

File: src/util.ts [hunk 11]
```diff
@@ -367,34 +371,30 @@ export function* flatMap<In, Out>(rng: Iterable<In>, fn: (item: In) => Iterable<
...
+            });
         }
-    }, {});
+    }
+    return converted_env;
 }
 
-export function normalizeEnvironmentVarname(varname: string) {
-    return process.platform === 'win32' ? varname.toUpperCase() : varname;
+export function fromDebuggerEnvironmentVars(debug_env?: DebuggerEnvironmentVariable[]): EnvironmentVariablesUndefined {
```
Comment (bobbrow) [inline src/util.ts:390]: If you take my feedback from `cmake-tools.ts`, this function should be removed.

File: test/backend-unit-tests/cmake/environment.test.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+            u: 'BBQ'
+        };
+        const resultA = EnvironmentVariablesUtils.mergeImpl(false, false, envA, undefined, envB);
+        expect(resultA).to.deep.equal({A: 'x', a: 'T', u: 'BBQ'});
+        const resultB = EnvironmentVariablesUtils.mergeImpl(true, false, envA, undefined, envB);
+        expect(resultB).to.deep.equal({A: 'x', B: null, a: 'T', u: 'BBQ'});
+        const resultC = EnvironmentVariablesUtils.mergeImpl(false, true, envA, undefined, envB);
+        expect(resultC).to.deep.equal({A: 'T', u: 'BBQ'});
+        const resultD = EnvironmentVariablesUtils.mergeImpl(true, true, envA, undefined, envB);
+        expect(resultD).to.deep.equal({A: 'T', B: null, u: 'BBQ'});
```
Comment (bobbrow) [inline test/backend-unit-tests/cmake/environment.test.ts:21]: Could you also add a test that has a `null` in envB override a value from envA? That's the presets case that I want to ensure doesn't break.

File: test/helpers/vscodefake/extensioncontext.ts [hunk 13]
```diff
@@ -38,8 +38,11 @@ export class DefaultExtensionContext implements vscode.ExtensionContext {
     get extensionMode(): vscode.ExtensionMode {
         throw new Error(notImplementedErr);
     }
+    extension: vscode.Extension<any>;
```
Comment (bobbrow) [inline test/helpers/vscodefake/extensioncontext.ts:41]: I don't understand what the changes in this file are for or how they're used.

**Review by bobbrow — CHANGES_REQUESTED** [review]
I still need to test it, but here is the initial review.

File: package.json [hunk 14]
```diff
@@ -21,7 +21,7 @@
     "native"
   ],
   "engines": {
-    "vscode": "^1.53.0"
+    "vscode": "^1.56.0"
```
[the remark this answers is not in the document]
Comment (lygstate) [reply package.json:24]: How about upgrade this first?

As the new vscode version have valid environment parameter

please refer https://github.com/microsoft/vscode-cmake-tools/pull/2236/commits/18accb7e121fe4fc177db76c6f732739cb0a9d16

File: src/config.ts [hunk 15]
```diff
@@ -23,9 +24,7 @@ export type StatusBarButtonVisibility = "default" | "compact" | "icon" | "hidden
 export type TouchBarButtonVisibility = "default" | "hidden";
 export type UseCMakePresets = 'always' | 'never' | 'auto';
 
-interface HardEnv {
-    [key: string]: string;
-}
+export type HardEnv = EnvironmentVariablesUndefined;
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/config.ts:27]: We can remove it in latter PR,

File: src/cmake-tools.ts [hunk 16]
```diff
@@ -2041,6 +2042,15 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
         return this.getCurrentLaunchTarget();
     }
 
+    async _getTargetLaunchEnvironment(drv: CMakeDriver | null, debug_env?: DebuggerEnvironmentVariable[]): Promise<EnvironmentVariablesUndefined> {
+        const env = util.fromDebuggerEnvironmentVars(debug_env);
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/cmake-tools.ts:2046]: Hi, this function is shared, the performance won't hurt by this, as we already using map

File: src/cmake-tools.ts [hunk 17]
```diff
@@ -2041,6 +2042,15 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+
+        // Add environment variables from ConfigureEnvironment.
+        const ctestEnv = await drv?.getConfigureEnvironment();
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/cmake-tools.ts:2049]: Hi, this doesn't change when it's presets, for non-presets, variants environment should include

File: src/cmake-tools.ts [hunk 18]
```diff
@@ -2151,22 +2158,9 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
-            });
-        }
-
-        // Add environment variables from configurePreset.
-        if (drv && this.configurePreset?.environment) {
-            const configure_preset_environment = await drv.getConfigureEnvironment();
-            launchEnv = util.mergeEnvironment(launchEnv, configure_preset_environment);
-        }
-
+        const launchEnv = await this._getTargetLaunchEnvironment(drv, user_config.environment);
```
Comment (lygstate) [inline src/cmake-tools.ts:2163]: @bobbrow This is the refactored out place.

File: src/cmake-tools.ts [hunk 19]
```diff
@@ -2101,15 +2111,12 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
         // Add debug configuration from settings.
         const user_config = this.workspaceContext.config.debugConfig;
         Object.assign(debug_config, user_config);
-        // Add environment variables from configurePreset.
-        if (this.configurePreset?.environment) {
-            const configure_preset_environment = await drv.getConfigureEnvironment();
-            debug_config.environment = debug_config.environment ? debug_config.environment.concat(util.makeDebuggerEnvironmentVars(configure_preset_environment)) : [];
-        }
-
+        const launchEnv = await this._getTargetLaunchEnvironment(drv, debug_config.environment);
```
Comment (lygstate) [inline src/cmake-tools.ts:2114]: @bobbrow This is the refactored out place.

File: src/environmentVariables.ts [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+export type EnvironmentVariablesItem = EnvironmentVariables | null | undefined;
+
+class EnvironmentVariablesPrivate {
+    private keyMapping: Map<string, string>;
+    public [envProperty]: EnvironmentVariables;
+    protected isWin32: boolean;
+    protected preserveNull: boolean;
+    constructor(
+        preserveNull: boolean | undefined,
+        isWin32: boolean | undefined) {
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/environmentVariables.ts:20]: Hi, this is placed here intensionlly, so that we can testing win32/non-win32 on windows and linux without
setup two operating system or waiting the github actions result.
That's simplify the test case a lot, we have no need remove it in the future.

File: src/environmentVariables.ts (hunk 1 above)
Comment (lygstate) [reply src/environmentVariables.ts:92]: > 
EnvironmentVariablesPrivate is proxied because we need
mantain compatiable with NodeJS.ProcessEnv.
For example, supporse we have a env named with `get`, if we using
typescript `Index Signatures`, then what's the result of env.get will have
two meaning:
  * call the function `get`
  * get the environment variable `get`
But for environment variable, access any member with `name` should return the expeted
environment variable for that  `name`

File: src/util.ts (hunk 11 above)
Comment (lygstate) [reply src/util.ts:390]: fromDebuggerEnvironmentVars is paired with makeDebuggerEnvironmentVars

File: src/drivers/driver.ts [hunk 21]
```diff
@@ -171,13 +172,13 @@ export abstract class CMakeDriver implements vscode.Disposable {
...
     /**
      * Compute the environment variables that apply with substitutions by expansionOptions
      */
-    async computeExpandedEnvironment(in_env: proc.EnvironmentVariables, expanded_env: proc.EnvironmentVariables): Promise<proc.EnvironmentVariables> {
-        const env = {} as { [key: string]: string };
+    async computeExpandedEnvironment(in_env: EnvironmentVariablesUndefined, expanded_env: EnvironmentVariablesUndefined): Promise<EnvironmentVariablesUndefined> {
+        const env = EnvironmentVariablesUtils.create();
         const opts = this.expansionOptions;
 
         await Promise.resolve(
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/drivers/driver.ts:184]: done

File: src/environmentVariables.ts [hunk 22]
```diff
@@ -0,0 +1,147 @@
+import * as os from 'os';
+import * as util from 'util';
+
+const inspect = util.inspect.custom;
+const envProperty = Symbol('envProperty');
+
+export type EnvironmentVariablesBasic = Record<string, string>;
+export type EnvironmentVariablesNull = Record<string, string | null>;
+export type EnvironmentVariablesUndefined = Record<string, string | undefined>; // alias of NodeJS.ProcessEnv
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/environmentVariables.ts:9]: > 
 Record<string, string | undefined> === Dict<string>

File: src/environmentVariables.ts [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+export type EnvironmentVariables = Record<string, string | undefined | null>;
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/environmentVariables.ts:10]: How to name EnvironmentVariablesUndefined?

File: src/environmentVariables.ts [hunk 24]
```diff
@@ -0,0 +1,147 @@
+import * as os from 'os';
+import * as util from 'util';
+
+const inspect = util.inspect.custom;
+const envProperty = Symbol('envProperty');
+
+export type EnvironmentVariablesBasic = Record<string, string>;
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/environmentVariables.ts:7]: Yeap, we can remove it

File: src/environmentVariables.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (lygstate) [reply src/environmentVariables.ts:10]: I wanna better name for
export type EnvironmentVariablesUndefined = Record<string, string | undefined>; // alias of NodeJS.ProcessEnv
export type EnvironmentVariables = Record<string, string | undefined | null>;
Name to Environment  and `EnvironmentWithNull`

File: src/environmentVariables.ts (hunk 3 above)
Comment (lygstate) [reply src/environmentVariables.ts:133]: It's called 11 times in preset, and maybe more in the future, that's worth another function

File: src/expand.ts (hunk 6 above)
Comment (lygstate) [reply src/expand.ts:100]: OK, the intension is do template like things, input string, return string, input null return null, that's we want

File: src/environmentVariables.ts [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+class EnvironmentVariablesPrivate {
+    private keyMapping: Map<string, string>;
+    public [envProperty]: EnvironmentVariables;
+    protected isWin32: boolean;
+    protected preserveNull: boolean;
+    constructor(
+        preserveNull: boolean | undefined,
+        isWin32: boolean | undefined) {
+        this.keyMapping = new Map<string, string>();
+        this[envProperty] = {};
```
[the remark this answers is not in the document]
Comment (lygstate) [reply src/environmentVariables.ts:37]: /* Using envProperty symbol is to provide valid implemention for [inspect]() */

File: test/helpers/vscodefake/extensioncontext.ts (hunk 13 above)
Comment (lygstate) [reply test/helpers/vscodefake/extensioncontext.ts:41]: Split to https://github.com/microsoft/vscode-cmake-tools/pull/2262

File: test/backend-unit-tests/cmake/environment.test.ts (hunk 12 above)
Comment (lygstate) [reply test/backend-unit-tests/cmake/environment.test.ts]: Added

File: src/environmentVariables.ts (hunk 3 above)
Comment (bobbrow) [reply src/environmentVariables.ts]: Ok, we can keep the one for presets.

**Review by bobbrow — APPROVED** [review]
Approved. After enabling the Windows CI #2295, I will merge with this, run one last test pass, and then commit.

**Comment by bobbrow** [comment]
One last question.  I still see a lot of direct interaction with `process.env` after this change.  Is the goal to make all of those accesses consistent too?  In other words, should we keep a global `Environment` that corresponds to `process.env` and enforce using that everywhere in the code?

**Comment by lygstate** [comment]
> process.env

Hi, there is no need modify `process.env`, I was using `Environment` to simulate typeof `process.env`, they indeed are the same thing after all.

**Comment by lygstate** [comment]
Can this be merged now?

**Comment by bobbrow** [comment]
@andreeis was almost done with the validation and then had to take some sick leave.  This will go in soon.  Thank you for your patience.

**Comment by bobbrow** [comment]
I merged it. We will continue to test as we have time, but I believe it will be good if future development already has this change in as it will add additional coverage.
